### PR TITLE
Tgui input list

### PR DESCRIPTION
**Pull request by ZeWaka**
[feat][qol]

## About the PR 

ports https://github.com/tgstation/tgstation/pull/56065, https://github.com/tgstation/tgstation/pull/56114
Licensed under MIT.

image

**Will need JS refactoring.**

## Why's this needed? 

pretty, more features:
* searchable
* multiple open

dreamseeker_2021-01-15_14-50-00
image

### CODE REVIEW

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                  for (index; index < buttons.length; index++) {
+                    if (buttons[index] === selectedButton) break;
+                  }
+                  index += direction;
+                  if (index < 0) index = buttons.length-1;
+                  else if (index >= buttons.length) index = 0;
+                  setSelectedButton(buttons[index]);
+                  setLastCharCode(null);
+                  document.getElementById(buttons[index]).focus();
+                  return;
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:83]:
```suggestion
                  let index = buttons.findIndex(selectedButton);
                  if (index === -1) {
                    index = 0;
                  }
                  index = (index + direction) % buttons.length;
                  const button = buttons[index];
                  setSelectedButton(button);
                  setLastCharCode(null);
                  document.getElementById(button).focus();
```
I know there's 100 ways to write this, but the one as-written has some *weird* patterns in it. Only thing that would need confirming is whether `findIndex` is available in the polyfill, as that's not in non-Edge IE.

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                  }
+                  else {
+                    foundValue = searchArray[0];
+                    setSearchIndex(0);
+                  }
+                }
+                else {
+                  const resultArray = displayedArray.filter(value =>
+                    value.substring(0, 1).toLowerCase() === charCode
+                  );
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:105]:
```suggestion
                  const resultArray = displayedArray.filter(value => (
                    value.substring(0, 1).toLowerCase() === charCode
                  ));
```

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                  <Flex.Item key={button} basis="100%">
+                    <Button
+                      color="transparent"
+                      content={button}
+                      id={button}
+                      width="100%"
+                      selected={selectedButton === button}
+                      onClick={() => {
+                        if (selectedButton === button) {
+                          act("choose", { choice: button });
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:146]:
```suggestion
                          act('choose', { choice: button });
```

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            </Section>
+          </Flex.Item>
+          {showSearchBar && (
+            <Flex.Item mt={1}>
+              <Input
+                fluid
+                onInput={(e, value) => setDisplayedArray(
+                  buttons.filter(val =>
+                    val.toLowerCase().search(value.toLowerCase()) !== -1
+                  )
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:171]:
```suggestion
                  buttons.filter(val => (
                    val.toLowerCase().search(value.toLowerCase()) !== -1
                  ))
```

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+          )}
+          <Flex.Item mt={1}>
+            <Flex textAlign="center">
+              <Flex.Item grow={1} basis={0}>
+                <Button
+                  fluid
+                  color="bad"
+                  lineHeight={2}
+                  content="Cancel"
+                  onClick={() => act("cancel")}
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:184]:
```suggestion
                  onClick={() => act('cancel')}
```

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                />
+              </Flex.Item>
+              <Flex.Item grow={1} basis={0} ml={1}>
+                <Button
+                  fluid
+                  color="good"
+                  lineHeight={2}
+                  content="Confirm"
+                  disabled={selectedButton === null}
+                  onClick={() => act("choose", { choice: selectedButton })}
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:194]:
```suggestion
                  onClick={() => act('choose', { choice: selectedButton })}
```

File: tgui/packages/tgui/interfaces/ListInput.js [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+              tabIndex={0}
+              onKeyDown={e => {
+                e.preventDefault();
+                if (lastScrollTime > performance.now()) {
+                  return;
+                }
+                lastScrollTime = performance.now() + 125;
+
+                if (e.keyCode === ARROW_KEY_UP || e.keyCode === ARROW_KEY_DOWN)
+                {
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/ListInput.js:69]:
```suggestion
                if (e.keyCode === ARROW_KEY_UP || e.keyCode === ARROW_KEY_DOWN) {
```

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]
Need to take a closer look at the JS re: how `lastCharCode` works in terms of how it affects the results. Little cautious with behavior with e.g. backspace.

File: code/modules/tgui/tgui_input_list.dm [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+	if (!user)
+		user = usr
+	if(!length(buttons))
+		return
+	if (!istype(user))
+		if (istype(user, /client))
+			var/client/client = user
+			user = client.mob
+		else
+			return
```
[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/tgui/tgui_input_list.dm:28]: If the passed argument isn't a mob or a client, it early returns.

[the remark this answers is not in the document]
Comment (mordent-goonstation) [reply code/modules/tgui/tgui_input_list.dm:28]: > If the passed argument isn't a mob or a client, it early returns.

Yes, but I was referring to the case of `user = client.mob` on line 26 being null for whatever reason. Unsure if a client can be mobless.

[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/tgui/tgui_input_list.dm:28]: Don't think it can.

[the remark this answers is not in the document]
Comment (ZeWaka) [reply code/modules/tgui/tgui_input_list.dm:28]: they can still be mobless while being disposed/GC'd

File: tgui/packages/tgui/interfaces/ListInput.js (hunk 1 above)
Comment (ZeWaka) [reply tgui/packages/tgui/interfaces/ListInput.js]: looks like it, at least in the bundle

Comment (ZeWaka) [reply tgui/packages/tgui/interfaces/ListInput.js]: didn't add the removal of `return`

**Comment by ZeWaka** [comment]
depending what gets merged first, this needs to be worked into a PR or into this PR: https://github.com/stylemistake/tgstation/blob/d12f1be3e7e1a2905462368490fecd080face51e/tgui/packages/tgui/interfaces/ListInput.js

File: code/modules/tgui/tgui_alert.dm [hunk 9]
```diff
@@ -99,7 +100,7 @@
  */
 /datum/tgui_modal/proc/wait()
 	while (!choice && !closed)
-		LAGCHECK(LAG_MED)
+		LAGCHECK(5)
```
Comment (ZeWaka) [inline code/modules/tgui/tgui_alert.dm:103]: Could be untenable as we don't have guarantees I don't think?

Comment (ZeWaka) [reply code/modules/tgui/tgui_alert.dm]: ok the byond internal scheduler might

**Pull request merged**